### PR TITLE
Add ability to add component to application via drag/drop

### PR DESCRIPTION
**Pull request by jeff-phillips-18**
https://jira.coreos.com/browse/ODC-840

cc @serenamarie125 @Veethika @sspeiche

### CODE REVIEW

File: frontend/packages/dev-console/src/components/topology/Topology.tsx [hunk 1]
```diff
@@ -35,6 +39,69 @@ export default class Topology extends React.Component<TopologyProps, State> {
...
+    const patch = [
+      {
+        op: 'replace',
+        path: '/metadata/labels',
+        value: labels,
+      },
+    ];
+
+    return k8sPatch(
+      dc.kind === 'DeploymentConfig' ? DeploymentConfigModel : DeploymentModel,
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/topology/Topology.tsx]: Move this logic to a utility.
We need to update ALL the resources related to this node and change the `part-of` label. Not just the DC itself.

File: frontend/packages/dev-console/src/components/topology/Topology.tsx [hunk 2]
```diff
@@ -35,6 +39,69 @@ export default class Topology extends React.Component<TopologyProps, State> {
...
+      dc,
+      patch,
+    );
+  };
+
+  onUpdateNodeGroup = (nodeId: string, targetGroup: string): void => {
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/topology/Topology.tsx]: Most of this function could be moved to a separate utility file.

File: frontend/packages/dev-console/src/components/topology/D3ForceDirectedRenderer.tsx [hunk 3]
```diff
@@ -425,22 +501,43 @@ export default class D3ForceDirectedRenderer extends React.Component<
...
+      targetGroup,
+    } = this.state;
     return (
-      <svg height={height} width={width} ref={this.refSvg} onClick={this.deselect}>
+      <svg
+        height={height}
+        width={width}
+        ref={this.refSvg}
+        onClick={this.deselect}
+        className={dragNodeId ? 'drag-active' : ''}
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/topology/D3ForceDirectedRenderer.tsx]: We should qualify this class name:
```suggestion
        className={dragNodeId ? 'odc-m-drag-active' : ''}
```
Or `odc-m-dragging`

File: frontend/packages/dev-console/src/components/topology/shapes/BaseNode.tsx [hunk 4]
```diff
@@ -81,37 +83,39 @@ export default class BaseNode extends React.Component<BaseNodeProps, State> {
...
-            </SvgBoxedText>
-          )}
-          {selected && (
-            <circle
-              className="odc-base-node__selection"
-              cx={0}
-              cy={0}
-              r={outerRadius + 1}
-              strokeWidth={2}
+          <g className={`odc-base-node__contents${isDragging ? ' is-highlight' : ''}`}>
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/topology/shapes/BaseNode.tsx]: We typically want to use the `classnames` library for joining classes.
This works fine though too.

File: frontend/packages/dev-console/src/components/topology/shapes/DefaultGroup.tsx [hunk 5]
```diff
@@ -20,42 +19,140 @@ function findLowestPoint<P extends Point>(points: P[]): P {
...
+    if (
+      point[0] < boundingBox[0][0] ||
+      point[0] > boundingBox[1][0] ||
+      point[1] < boundingBox[0][1] ||
+      point[1] > boundingBox[1][1]
+    ) {
+      return false;
+    }
+
+    return d3.polygonContains(borderPoints, point);
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/topology/shapes/DefaultGroup.tsx]: Can we use this library instead for more precision?
`https://github.com/maxnachlinger/point-in-svg-path`

File: frontend/packages/dev-console/src/components/topology/D3ForceDirectedRenderer.tsx [hunk 6]
```diff
@@ -516,6 +615,7 @@ type GroupWrapperProps = GroupProps & {
   component: React.ComponentType<GroupProps>;
   onEnter(GroupSelection): void;
   view: ViewGroup;
+  groupRef: any;
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/topology/D3ForceDirectedRenderer.tsx]: Give this a proper type.

File: frontend/packages/dev-console/src/components/topology/D3ForceDirectedRenderer.tsx [hunk 7]
```diff
@@ -535,7 +635,7 @@ class GroupWrapper extends React.Component<GroupWrapperProps> {
   }
 
   render() {
-    const { component: Component, onEnter, view, ...other } = this.props;
-    return <Component {...other} />;
+    const { component: Component, onEnter, view, groupRef, ...other } = this.props;
+    return <Component groupRef={groupRef} {...other} />;
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/topology/D3ForceDirectedRenderer.tsx]: Shouldn't `groupRef` should just be ref and all groups should provide an api?

File: frontend/packages/dev-console/src/components/topology/shapes/DefaultGroup.tsx [hunk 8]
```diff
@@ -20,42 +21,90 @@ function findLowestPoint<P extends Point>(points: P[]): P {
...
-  if (nodes.length === 0) {
-    return null;
-  }
+interface DefaultGroupState {
+  nodes: ViewNode[];
+  lowestPoint: PointWithSize;
+  containerPath: string;
+}
 
+const getUpdatedStateValues = (name: string, nodes: ViewNode[]) => {
```
Comment (sahil143) [inline frontend/packages/dev-console/src/components/topology/shapes/DefaultGroup.tsx]: name is not being used in the method.

**Comment by jeff-phillips-18** [comment]
@sahil143 updated per your comments, PTAL

**Comment by jeff-phillips-18** [comment]
@sbose78

**Comment by sbose78** [comment]
deployed the PR to https://pr-2141-openshift-console.apps.mydevaws1.devcluster.openshift.com/ ( credentials in # devconsole-ui

**Comment by sbose78** [comment]
The drag-n-drop works really well, @jeff-phillips-18 !
One tiny feedback, even though datagrid-service has been brought out of the app group, it is still half-present inside it in the visualization.

image

**Comment by sbose78** [comment]
Curious why do we have a confirmation box when we wish to take a component out of the application but not when we are bringing a component into an application ? :)

**Comment by jeff-phillips-18** [comment]
@sbose78 Design doc is at https://docs.google.com/document/d/1mIld9b_mC5sGZEvHHfwflQpFZVcrrwG029egE3550kg/edit?ts=5d540c70#heading=h.jwcv3fh2rvtn

Had spoken with UX and the confirmation is only needed when the node is already part of an application (move to another or remove from the application).

**Comment by sbose78** [comment]
Thanks @jeff-phillips-18 !

**Comment by sspeiche** [comment]
@sbose78 there are cases where we determined it made sense to not to (it becomes annoying) and other cases where it would be good to clarify (prompt). Feel free to add your thoughts in the design. We'd like the get the current design in user's hands and get some feedback on it.

**Comment by sbose78** [comment]
Sounds reasonable, Steve.

**Comment by jeff-phillips-18** [comment]
Modified code to update the 'app.kubernetes.io/part-of' label on resources which have the same 'app.kubernetes.io/instance' and 'app.kubernetes.io/part-of' labels.

File: frontend/packages/console-shared/src/utils/application-utils.ts [hunk 9]
```diff
@@ -0,0 +1,77 @@
+import { k8sList, k8sPatch, K8sResourceKind, modelFor } from '@console/internal/module/k8s';
+import * as _ from 'lodash';
+
+const updateItemAppLabel = (item: K8sResourceKind, application) => {
```
Comment (christianvogt) [inline frontend/packages/console-shared/src/utils/application-utils.ts]: Missing param and return types.
Shared code would be good to have a comment describing the purpose.

File: frontend/packages/console-shared/src/utils/application-utils.ts [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+  const labels = _.cloneDeep(item.metadata.labels || {});
```
Comment (christianvogt) [inline frontend/packages/console-shared/src/utils/application-utils.ts]: Could this be:
```
  const labels = { ...item.metadata.labels, 'app.kubernetes.io/part-of': application || undefined };
```

If you still need to clone, it should be done after line 10.

File: frontend/packages/console-shared/src/utils/application-utils.ts [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    return Promise.all(patches);
+  }
+
+  // Retrieve all resources with the same instance label
+  const kinds = ['ReplicationController', 'Route', 'Service', 'ReplicaSet', 'BuildConfig', 'Build'];
+  _.forEach(kinds, (kind) => {
+    lists.push(
+      k8sList(modelFor(kind), {
+        ns: resource.metadata.namespace,
+        labelSelector: { 'app.kubernetes.io/instance': instanceName },
```
Comment (christianvogt) [inline frontend/packages/console-shared/src/utils/application-utils.ts]: Can you select on multiple labels to include `part-of`?

File: frontend/packages/console-shared/src/utils/application-utils.ts [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+  const lists: Promise<any>[] = [];
+  const patches: Promise<any>[] = [updateItemAppLabel(resource, application)];
+
+  // If there is no instance label, only update this item
+  if (!instanceName) {
+    return Promise.all(patches);
+  }
+
+  // Retrieve all resources with the same instance label
+  const kinds = ['ReplicationController', 'Route', 'Service', 'ReplicaSet', 'BuildConfig', 'Build'];
```
Comment (christianvogt) [inline frontend/packages/console-shared/src/utils/application-utils.ts]: Do we need to list kinds? What will happen with knative resources?

File: frontend/packages/dev-console/src/utils/svg-utils.ts [hunk 13]
```diff
@@ -31,6 +31,12 @@ function roundedHull1(polyPoints: Point[], hp: HullPaddingGetter): string {
   return `M ${p1} A ${padding},${padding},0,0,0,${p2} A ${padding},${padding},0,0,0,${p1}`;
 }
 
+// Return a point that extends a segment a given length
+export function extensionPoint(a: Point, b: Point, length: number): Point {
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/utils/svg-utils.ts]: This doesn't look to be used anywhere.

Comment (jeff-phillips-18) [reply frontend/packages/dev-console/src/utils/svg-utils.ts]: yup, left over (but used in a dependent PR). I'll remove and add to that PR.

File: frontend/packages/console-shared/src/utils/application-utils.ts (hunk 12 above)
Comment (jeff-phillips-18) [reply frontend/packages/console-shared/src/utils/application-utils.ts]: Each `kind` needs to be retrieved separately. From what I can tell, there are no knative resources that use the 'app.kubernetes.io/part-of' label.

Comment (christianvogt) [reply frontend/packages/console-shared/src/utils/application-utils.ts]: @invincibleJai do serverless resources adhere to the labeling conventions? You know the labels we apply at import time. And if so, does the controller propagate those labels to the child resources it creates?

**Pull request merged**